### ISSUE TITLE
ACF field values not loaded correct in new widget block editor

### ISSUE THREAD

**Issue by christiaanstijnen** [AdvancedCustomFields/acf#611]
### Description
In the new widget block editor we have the problem that if ACF fields are used in the widget form the field values of the wrong instance might be loaded.

### Setup
Tested with a clean install of WP 5.9, Theme: Twenty Twenty-One

### Scenario.
1. Create a custom widget by extending the WP_Widget class. (https://developer.wordpress.org/themes/functionality/widgets/)
2. Add an ACF field to the widget form. (i.e. a text field)
3. Insert the custom widget in a sidebar and insert text ‘first widget’.
4. Add another instance of this custom widget after the first instance in the sidebar and insert text ‘second widget’.
5. Save the sidebar.
6. Refresh the page
7. Now switch positions of the widgets, so the widget named second widget is in the 1st place of appearance
8. Save the sidebar
9. Refresh te page
10. Notice the content of the widgets. The 1st instance has ‘first widget’ again and the 2nd instance has ’second widget’ again

### My findings

The new Gutenberg editor passes a number in the async request which loads the field values for the widget instance. This number seems like a simple increment of the position the widget is in. But this number is treated like it is the unique ID for the widget. This is why the data is getting mixed up.

**Comment by lgladdy** [comment]
Hey @christiaanstijnen, thanks for the report! We'll take a look at this and get back to you shortly.

**Comment by lgladdy** [comment]
@christiaanstijnen Could you share your widget class and any other related code so we can replicate your exact setup please? Feel free to email it in if you don't want to share on here!

**Comment by christiaanstijnen** [comment]
@lgladdy thanks for your reply, I added my test code in the opening post. You do need to create the acf fields via ACF your self, is that ok?

Create a field group for 'Test Widget' with a text field in it named 'test_text'

**Comment by lgladdy** [comment]
@christiaanstijnen Thanks for that. It's looking like WordPress's legacy widget block (which is what powers legacy widgets in the block editor) won't support ACF fields because of the way it generates IDs. It resets the IDs whenever you move things around or delete them, so there's no way for us to know what fields should actually be accessed, and rather than a widget having a unique ID, it's now just a simple number.

I think the answer to this will be to either disable the new widget block editor (`remove_theme_support( 'widgets-block-editor' );`) and use the legacy system entirely, or upgrade your legacy widgets to ACF Blocks in ACF PRO, which store the data against the widget correctly.

**Comment by lgladdy** [comment]
@christiaanstijnen Yeh, it's pretty weird to be honest! I'm still looking into it, because it does look like when WordPress tries to render the widget, it is sending the correct field values over - but obviously ACF isn't expecting them in that payload - so by calling `get_field('name', 'widget_'.$id)` it won't be found there (and indeed, looking at my 3 widgets, somehow WordPress is sending the same widget ID for all 3 instances in the sidebar...)

That said, if you `var_dump($instance)` in your widget function, you'll see that the field value is in there - so you could get the data that way instead if you wanted, but given that would still require widget updates, it might just be better to switch to blocks.

What's also weird, is widget previews in the widget block editor render with the wrong IDs anyway:
![Screenshot 2022-02-17 at 11 19 12](https://user-images.githubusercontent.com/213440/154471118-c945079f-7a4b-4ff3-9c59-4186de425ca1.png)

But on the front end, they're "correct", but still just simple IDs that reorder as you move around, without updating the other fields:
![Screenshot 2022-02-17 at 11 19 20](https://user-images.githubusercontent.com/213440/154471197-9e9a441f-f6ee-42e6-b0e3-3e37ce7587d0.png)

**Comment by christiaanstijnen** [comment]
Yeah it the widget types controller you can see that on rendering the widget for the first time in the editor it uses the number to 'set' the widget. That number I think ACF uses that number to get the option.

https://github.com/WordPress/wordpress-develop/blob/5.9/src/wp-includes/rest-api/endpoints/class-wp-rest-widget-types-controller.php#L467

there is a filter 'widget_update_callback' where several instances and data of the widget are available. Perhaps there could be an opportunity to fix the field values.
